ansible_runner/streaming.py

```diff
@@ -81,6 +89,64 @@ def __init__(self, _input=None, _output=None, **kwargs):
         self.status = "unstarted"
         self.rc = None
 
+    def _begin_keepalive(self):
+        """Starts a keepalive thread at most once"""
+        if not self._keepalive_thread:
+            self._keepalive_thread = Thread(target=self._keepalive_loop)
+            self._keepalive_thread.start()
+
+    def _end_keepalive(self):
+        """Disable the keepalive interval and notify the keepalive thread to shut down"""
+        self._keepalive_interval_sec = 0
+        self._output_event.set()
+
+    def _keepalive_loop(self):
+        """Main loop for keepalive injection thread; exits when keepalive interval is <= 0"""
+        while self._keepalive_interval_sec > 0:
+            # block until output has occurred or keepalive interval elapses
+            if self._output_event.wait(timeout=self._keepalive_interval_sec):
+                # output was sent before keepalive timeout; reset the event and start waiting again
+                self._output_event.clear()
+                continue
+
+            # keepalive interval elapsed; try to send a keepalive...
+            # pre-acquire the output lock without blocking
+            if not self._output_lock.acquire(blocking=False):
```

Comment:
Just FYI: This will never be false because `_output_lock` is a `RLock` which is reentrant (not "read" as a surprising number of developers seem to think!) e.g.,
```python
>>> from threading import RLock, Event
>>> thing = Event()
>>> lock = RLock()
>>> lock.acquire(blocking=False)
True
>>> lock.acquire(blocking=False)
True
>>> lock.acquire(blocking=False)
True
>>> lock.acquire(blocking=False)
True
>>> lock.
lock.acquire(  lock.release(  
>>> lock.release()
>>> lock.release()
>>> lock.release()
>>> lock.release()
>>> lock.release()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: cannot release un-acquired lock
```
You can grab it all you want but it's never released until you call `release()` the same number of times you called `acquire()`. It's meant for preventing deadlock by letting the same thread grab a lock more than once or for multiple threads to hold a lock.

Reply:
I think you might be misunderstanding what's going on in that method (remember it's running on a dedicated thread)- I specifically chose an `RLock` to allow it to be "best effort" pre-acquired by the thread, then "inherited" by the call to the event writer from the same thread (which will re-entrantly reacquire the lock). If something else has picked up the lock, it means the main thread is in the middle of writing an event, in which case the keepalive is redundant and should be skipped. Without this lightweight pre-acquisition + reentrant property of RLock, the keepalive thread would block on the event dispatch until the main thread released the lock and immediately splat a redundant keepalive (which in some cases, like during shutdown, could also technically break the output by emitting a keepalive after the `eof` event).

Reply:
I think I see what you mean. It's somewhat recursive ("inherited," "same thread"). That makes sense. I don't know enough about the codebase to say any more about that. However, I think my other comment is still valid, right? This conditional on 114 will never be false. I also don't think it will ever block even if `blocking=True`. Thoughts on that?

Reply:
Nope, not when the main thread already holds the lock, which is the case I'm trying to correct for:

* (on main thread) idle for a long time
* (on keepalive thread) keepalive timeout expires
* (on main thread) acquire lock, start event write
* (on keepalive thread) acquire lock to start a keepalive write, block because the lock is held
* (on main thread) complete event write, release lock
* (on keepalive thread) acquire lock, write superfluous keepalive

this fixes that problem, because the extra non-blocking lock acquire on the keepalive thread will return `False` (since the main thread holds the lock), so we know there's another write already in progress and that we don't need a keepalive

Reply:
(if you don't believe me, set the keepalive interval to 1s and inject a 5s delay in one of the event writers, which is what I did when I was verifying it :laughing: )

Reply:
Thanks for the explanation! It's clarifying. I follow what you're saying, but I think I would have to see it running, like you suggested, to really understand the code. I'm going to try to do that, but I wouldn't hold up a PR for it (I'm new and still need to learn some basics). Sounds like you've tested it, and it works the way you expect!

Reply:
Oh, I see the main idea I was missing: Python's RLock.acquire() does return False if a *different* thread tries to acquire. Cool.

Reply:
@jjwatt if you want to see how I set that up to observe it in a debugger sometime, give me a poke on Slack...